tests/integration/test_charm_with_loki.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import json
+import time
+
+import pytest
+import requests
+from charms.loki_k8s.v0.loki_push_api import (
+    BINARY_FILE_NAME,
+    WORKLOAD_BINARY_DIR,
+    WORKLOAD_CONFIG_DIR,
+    WORKLOAD_CONFIG_FILE_NAME,
+    WORKLOAD_CONFIG_PATH,
+)
+from helpers import all_combinations
+from pytest_operator.plugin import OpsTest
+
+# copied over from log.py to avoid circular imports
+TEST_JOB_NAME = "test-job"
+SYSLOG_LOG_MSG = "LOG SYSLOG"
+LOKI_LOG_MSG = "LOG LOKI"
+FILE_LOG_MSG = "LOG FILE"
+
+
+async def run_in_loki_tester(ops_test: OpsTest, cmd):
+    cmd_line = f"juju ssh --container loki-tester loki-tester/0 {cmd}"
+    return await ops_test.run(*cmd_line.split(" "))
```

Comment:
what functionality are you referring to? it's just a utility function I use when I need to do stuff in the workload container that I found hard/impossible to do via pytest_operator.

Reply:
mmh I disagree, or maybe I don't understand your point. So trying to be specific: let's look at `test_promtail_is_running_in_workload_container`. This test verifies that, as a consequence of using LogProxyConsumer in the tester charm, promtail is running in the workload container. We are testing LogProxyConsumer, not the tester. We are testing that LogProxyConsumer does its job and that in the workload container there is a promtail process running. If it were not running, the subsequent tests about syslog and file tailing would also fail, because they depend on promtail being there.

Do you think that this test is not useful? My idea was to have a test for the precondition (is promtail running) and then tests for the consequences (does promtail actually work and send stuff to loki). So that we can detect failures early on and not just 'loki isn't receiving stuff', but 'promtail isn't up', which is easier debugging.

Reply:
I think this makes sense to have.